### PR TITLE
Add truncation to SFT DataCollatorForLanguageModeling

### PR DESCRIPTION
**Pull request by albertvillanova**
Add truncation to SFT `DataCollatorForLanguageModeling`.

This PR adds support for sequence truncation to the `DataCollatorForLanguageModeling` class, allowing sequences to be truncated to a specified maximum length from either the start or end. It also introduces comprehensive tests to ensure correct behavior for different truncation modes and edge cases.

This PR aligns SFT `DataCollatorForLanguageModeling` with the existing SFT `DataCollatorForVisionLanguageModeling`, which already truncates inputs.

Follow-up to:
- #5306

### Changes

Enhancements to sequence truncation:

* Added `max_length` and `truncation_mode` parameters to `DataCollatorForLanguageModeling`, enabling truncation of sequences longer than `max_length` with options to keep either the start or end tokens.
* Implemented per-sequence truncation logic in the `torch_call` method, applying the specified truncation mode and ensuring associated masks (such as `completion_mask` and `assistant_masks`) are truncated consistently. A `ValueError` is raised for unsupported truncation modes.

Testing improvements:

* Added multiple test cases covering:
  - Truncation from the start and end.
  - Behavior when no truncation is needed.
  - Truncation with completion masks.
  - Handling of invalid truncation modes.

---

> [!NOTE]
> **Medium Risk**
> Changes how SFT batches are built by introducing truncation in `DataCollatorForLanguageModeling`, which can alter training inputs/labels and raise new configuration errors (especially around `padding_free` and dataset prep). Test coverage mitigates but behavior changes may impact existing training setups relying on untruncated batches.
> 
> **Overview**
> Adds optional per-example truncation to `DataCollatorForLanguageModeling` via new `max_length` and `truncation_mode` (`keep_start`/`keep_end`), ensuring `labels`, `completion_mask`, and `assistant_masks` are truncated consistently and rejecting unsupported modes.
> 
> Wires `SFTTrainer` to pass `max_length`/`truncation_mode` into the text collator (disabled when `padding_free=True`) and introduces a guard that errors when `skip_prepare_dataset=True` is used with `padding_free=True` and a non-`None` `max_length`.
> 
> Extends tests to cover truncation modes, mask interactions, invalid modes, and trainer configuration propagation/validation.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 281c8156d5e90858f13898489a4d57ac965627a9. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Comment by qgallouedec** [comment]
Is the next step to remove https://github.com/albertvillanova/trl/blob/c3a81dea136d18c607813211f7156fbc720aecd1/trl/trainer/sft_trainer.py#L1209-L1211?

**Comment by albertvillanova** [comment]
> Is the next step to remove...

Yes @qgallouedec, analogously to what I already proposed in the precedent PR (see motivation 3 in https://github.com/huggingface/trl/pull/5305#issuecomment-4083192554), my refactoring plan is to remove the truncation after the padding, present in the body of the trainer; and require it to be done by the collator.

File: trl/trainer/sft_trainer.py [hunk 1]
```diff
@@ -894,6 +920,12 @@ def __init__(
             and args.dataset_kwargs.get("skip_prepare_dataset", False)
             or self._is_vision_dataset
         )
+        if skip_prepare_dataset and self.padding_free and args.max_length is not None and not self._is_vision_dataset:
+            raise ValueError(
+                "When `padding_free=True`, `max_length` must be enforced during dataset preparation or packing, not in "
+                "the collator. Disable `skip_prepare_dataset`, provide already packed/truncated inputs, or set "
+                "`max_length=None`."
```
Comment (qgallouedec) [inline trl/trainer/sft_trainer.py:927]: Ok so you're saying: if you decide to skip the dataset preparation, then padding-free isn't supported with max_length, because we would need to truncate the `seq_lengths` column, which is quite annoying to do

Comment (qgallouedec) [reply trl/trainer/sft_trainer.py:927]: If my understanding is correct, then I agree. Fortunately, `args.dataset_kwargs.get("skip_prepare_dataset") is True` is very rare as far as I know

File: trl/trainer/sft_trainer.py [hunk 2]
```diff
@@ -164,26 +172,42 @@ class DataCollatorForLanguageModeling(DataCollatorMixin):
...
+            completion_mask = [torch.tensor(example["completion_mask"]) for example in examples]
+        if "assistant_masks" in examples[0]:
+            assistant_masks = [torch.tensor(example["assistant_masks"]) for example in examples]
+
+        # Truncate per sequence if necessary
+        if self.max_length is not None and not self.padding_free:
+            if self.truncation_mode == "keep_start":
+                sl = slice(None, self.max_length)
+            elif self.truncation_mode == "keep_end":
+                sl = slice(-self.max_length, None)
```
Comment (qgallouedec) [inline trl/trainer/sft_trainer.py:190]: nice!
